pybids 0.6.4 causes test failure in BIDSDataGrabber

### Summary

The error in https://travis-ci.org/nipy/nipype/builds/408021599:

```
______________________________ test_bids_grabber _______________________________
[gw0] linux -- Python 3.5.5 /home/travis/virtualenv/python3.5.5/bin/python
tmpdir = local('/tmp/pytest-of-travis/pytest-0/popen-gw0/test_bids_grabber0')
    @pytest.mark.skipif(not have_pybids,
                        reason="Pybids is not installed")
    @pytest.mark.skipif(sys.version_info < (3, 0),
                        reason="Pybids no longer supports Python 2")
    @pytest.mark.skipif(not dist_is_editable('pybids'),
                        reason="Pybids is not installed in editable mode")
    def test_bids_grabber(tmpdir):
        tmpdir.chdir()
        bg = nio.BIDSDataGrabber()
        bg.inputs.base_dir = os.path.join(datadir, 'ds005')
        bg.inputs.subject = '01'
        results = bg.run()
        assert os.path.basename(results.outputs.anat[0]) == 'sub-01_T1w.nii.gz'
>       assert os.path.basename(results.outputs.func[0]) == (
            'sub-01_task-mixedgamblestask_run-01_bold.nii.gz')
E       AssertionError: assert 'sub-01_task-...01_events.tsv' == 'sub-01_task-m...1_bold.nii.gz'
E         - sub-01_task-mixedgamblestask_run-01_events.tsv
E         ?                                     ^^^ ^^ ^^^
E         + sub-01_task-mixedgamblestask_run-01_bold.nii.gz
E         ?                                     ^^^^^ ^^ ^^
/home/travis/build/nipy/nipype/nipype/interfaces/tests/test_io.py:596: AssertionError
```

@adelavega @tyarkoni Flagging you in this in case the cause is obvious.

It's not immediately clear whether the correct fix will be in pybids or nipype.

### Execution environment

Travis CI

--- a/nipype/interfaces/io.py
+++ b/nipype/interfaces/io.py
@@ -2828,7 +2828,8 @@ class BIDSDataGrabber(IOBase):
         return runtime
 
     def _list_outputs(self):
-        layout = gb.BIDSLayout(self.inputs.base_dir)
+        layout = gb.BIDSLayout(self.inputs.base_dir,
+                               exclude=['derivatives/', 'code/', 'sourcedata/'])
 
         # If infield is not given nm input value, silently ignore
         filters = {}

--- a/nipype/interfaces/io.py
+++ b/nipype/interfaces/io.py
@@ -2804,8 +2804,8 @@ class BIDSDataGrabber(IOBase):
 
         if not isdefined(self.inputs.output_query):
             self.inputs.output_query = {
-                "func": {"modality": "func", extensions=['nii', '.nii.gz']},
-                "anat": {"modality": "anat", extensions=['nii', '.nii.gz']},
+                "func": {"modality": "func", 'extensions': ['nii', '.nii.gz']},
+                "anat": {"modality": "anat", 'extensions': ['nii', '.nii.gz']},
                 }
 
         # If infields is empty, use all BIDS entities

--- a/nipype/interfaces/io.py
+++ b/nipype/interfaces/io.py
@@ -2803,8 +2803,10 @@ class BIDSDataGrabber(IOBase):
         super(BIDSDataGrabber, self).__init__(**kwargs)
 
         if not isdefined(self.inputs.output_query):
-            self.inputs.output_query = {"func": {"modality": "func"},
-                                        "anat": {"modality": "anat"}}
+            self.inputs.output_query = {
+                "func": {"modality": "func", extensions=['nii', '.nii.gz']},
+                "anat": {"modality": "anat", extensions=['nii', '.nii.gz']},
+                }
 
         # If infields is empty, use all BIDS entities
         if infields is None and have_pybids:

--- a/nipype/interfaces/io.py
+++ b/nipype/interfaces/io.py
@@ -2755,6 +2755,8 @@ class BIDSDataGrabberInputSpec(DynamicTraitedSpec):
                                  desc='Generate exception if list is empty '
                                  'for a given field')
     return_type = traits.Enum('file', 'namedtuple', usedefault=True)
+    strict = traits.Bool(desc='Return only BIDS "proper" files (e.g., '
+                              'ignore derivatives/, sourcedata/, etc.)')
 
 
 class BIDSDataGrabber(IOBase):
@@ -2828,8 +2830,10 @@ class BIDSDataGrabber(IOBase):
         return runtime
 
     def _list_outputs(self):
-        layout = gb.BIDSLayout(self.inputs.base_dir,
-                               exclude=['derivatives/', 'code/', 'sourcedata/'])
+        exclude = None
+        if self.inputs.strict:
+            exclude = ['derivatives/', 'code/', 'sourcedata/']
+        layout = gb.BIDSLayout(self.inputs.base_dir, exclude=exclude)
 
         # If infield is not given nm input value, silently ignore
         filters = {}

--- a/nipype/interfaces/tests/test_io.py
+++ b/nipype/interfaces/tests/test_io.py
@@ -592,9 +592,9 @@ def test_bids_grabber(tmpdir):
     bg.inputs.base_dir = os.path.join(datadir, 'ds005')
     bg.inputs.subject = '01'
     results = bg.run()
-    assert os.path.basename(results.outputs.anat[0]) == 'sub-01_T1w.nii.gz'
-    assert os.path.basename(results.outputs.func[0]) == (
-        'sub-01_task-mixedgamblestask_run-01_bold.nii.gz')
+    assert 'sub-01_T1w.nii.gz' in map(os.path.basename, results.outputs.anat)
+    assert 'sub-01_task-mixedgamblestask_run-01_bold.nii.gz' in \
+        map(os.path.basename, results.outputs.func)
 
 
 @pytest.mark.skipif(not have_pybids,
@@ -610,7 +610,7 @@ def test_bids_fields(tmpdir):
     bg.inputs.subject = '01'
     bg.inputs.output_query['dwi'] = dict(modality='dwi')
     results = bg.run()
-    assert os.path.basename(results.outputs.dwi[0]) == 'sub-01_dwi.nii.gz'
+    assert 'sub-01_dwi.nii.gz' in map(os.path.basename, results.outputs.dwi)
 
 
 @pytest.mark.skipif(not have_pybids,

